### DISCUSSION TITLE
gRPC Transport Speed Test

### DISCUSSION THREAD

**Discussion by changyp6** [XTLS/Xray-core#381] [Performance]
Just did a quick test on gRPC transport
I tested 2 transports, gPPC and Websocket
```
gPRC Transport :     "VLESS + gPRC + CloudFlare CDN" --> Nginx 443 port --> Inbound VLESS + gPRC (UNIX Domain Socket)
Websocket Transport: "VLESS +   WS + CloudFalre CDN" --> Nginx 443 port --> Inbound VLESS +   WS (UNIX Domain Socket)
```
Test site is www.speedtest.net

Test result is
```
              Tun:   upload / download                MultiTun: upload / download
gRPC:      6Mbits ~ 8Mbits / 6Mbits ~ 8Mbits        7Mbits ~ 9Mbits / 6Mbits ~ 8Mbits

             with ?ed=2048    upload / download          without ?ed=2048   upload  / download
Websocket:            19Mbits / 19Mbits                            19Mbits / 19Mbits
```

It seems that gPRC transport will make significant transport speed drop

I'm not sure if it's my settings issue which make the speed drop, or it is just gRPC's speed limit.

**Comment by ghost** [comment]
1. Speedtest.net is NOT a proper tool for measurement.
2. Using gRPC may impact performance in bad internet condition.
3. It’s can’t be reproduced in testing.
4. gRPC(multi) has better performance than WebSocket in single-thread testing, and much quicker than Mux.Cool in multi-threads testing.

**Reply by changyp6** [reply]
What is the PROPER tool for speed measurement ?

> 1. Speedtest.net is NOT a proper tool for measurement.
> 2. Using gRPC may impact performance in bad internet condition.
> 3. It’s can’t be reproduced in testing.
> 4. gRPC(multi) has better performance than WebSocket in single-thread testing, and much quicker than Mux.Cool in multi-threads testing.

**Reply by ghost** [reply]
iperf3

**Reply by changyp6** [reply]
The real case is when I use gPRC transport to watch youtube video, the auto resolution is 480p, when I switch back to websocket, the auto resolution is 1080p

I'm not sure if gPRC transport speed is limited by Cloudflare, but this is the result I got from practice.
> iperf3

**Reply by bestmikecat** [reply]
I tried it and the result is the same as yours, it's really slow in practice, so I still use websocket.

**Comment by unknowedNp** [comment]
> Just did a quick test on gRPC transport I tested 2 transports, gPPC and Websocket
> 
> ```
> gPRC Transport :     "VLESS + gPRC + CloudFlare CDN" --> Nginx 443 port --> Inbound VLESS + gPRC (UNIX Domain Socket)
> Websocket Transport: "VLESS +   WS + CloudFalre CDN" --> Nginx 443 port --> Inbound VLESS +   WS (UNIX Domain Socket)
> ```
> 
> Test site is [www.speedtest.net](http://www.speedtest.net)
> 
> Test result is
> 
> ```
>               Tun:   upload / download                MultiTun: upload / download
> gRPC:      6Mbits ~ 8Mbits / 6Mbits ~ 8Mbits        7Mbits ~ 9Mbits / 6Mbits ~ 8Mbits
> 
>              with ?ed=2048    upload / download          without ?ed=2048   upload  / download
> Websocket:            19Mbits / 19Mbits                            19Mbits / 19Mbits
> ```
> 
> It seems that gPRC transport will make significant transport speed drop
> 
> I'm not sure if it's my settings issue which make the speed drop, or it is just gRPC's speed limit.

Can you share your config? I config nginx without cdn that ok and connect but when use CDN, not working well

**Comment by iopq** [comment]
Easy test to do: use Trojan gRPC vs. xray

Trojan is faster for some reason

edit: it is sometimes faster
